### ISSUE TITLE
DataFetcher and convolutionFlat vs ImageRecordReader and convolution

### ISSUE THREAD

**Issue by kobaeugenea** [deeplearning4j/deeplearning4j#3489]
Hello! Sorry for my bad english. I compare two similar network model, but it not give expected result.
`
    public static void main(String[] args) throws Exception {

        int outputNum = 106; // number of output classes
        int nChannels = 1; // Number of input channels
        int batchSize = 64; // Test batch size
        int iterations = 1; // Number of training iterations
        int seed = 123; //

        DataSetIterator trafficSignTrainer = new BaseDatasetIterator(batchSize, 70687, new TrafficSignDataFetcher("rtsd-r3/gt_train.csv",
                                                                                                                  "rtsd-r3/train"));
        DataSetIterator trafficSignTest = new BaseDatasetIterator(batchSize, 22967, new TrafficSignDataFetcher("rtsd-r3/gt_test.csv",
                                                                                                              "rtsd-r3/test"));
        log.info("Build model....");
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations) // Training iterations as above
                .regularization(true).l2(0.0005)
                .learningRate(.01)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .list()
                .layer(0, new ConvolutionLayer.Builder(5, 5)
                        .nIn(nChannels)
                        .stride(1, 1)
                        .nOut(20)
                        .activation(Activation.IDENTITY)
                        .build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(2, new ConvolutionLayer.Builder(5, 5)
                        //Note that nIn need not be specified in later layers
                        .stride(1, 1)
                        .nOut(50)
                        .activation(Activation.IDENTITY)
                        .build())
                .layer(3, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(4, new DenseLayer.Builder().activation(Activation.RELU)
                                                  .nOut(500).build())
                .layer(5, new OutputLayer.Builder(LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nOut(outputNum)
                        .activation(Activation.SOFTMAX)
                        .build())
                .setInputType(InputType.convolutionalFlat(28, 28, 1)) //See note below
                .backprop(true).pretrain(false).build();

        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();
        log.info("Train model....");
            model.fit(trafficSignTrainer);
            Evaluation eval = new Evaluation(outputNum);
            while(trafficSignTest.hasNext()){
                DataSet ds = trafficSignTest.next();
                INDArray output = model.output(ds.getFeatureMatrix(), false);
                eval.eval(ds.getLabels(), output);

            }
            log.info(eval.stats());
            trafficSignTest.reset();
        }
`
Where TrafficSignDataFetcher - is my DataFetcher, it normalize pixel to range from 0 to 1, resize image to size 28x28px. This case give me result:
 Accuracy:        0,6823
 Precision:       0,6227
 Recall:          0,3446
 F1 Score:        0,4437
But when I use ImageRecordReader I get result:
  Accuracy:        0,0251
 Precision:       0,0914
 Recall:          0,0751
 F1 Score:        0,0824
   In both case I use same datasets.
Code for second case:
`
public static void main(String[] args) throws Exception {
        int height = 28;
        int width = 28;
        int channels = 1;
        int numLabels = 106;
        int batchSize = 64;

        long seed = 123;
        Random rng = new Random(seed);
        int iterations = 1;

        log.info("Load data....");
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();

        File mainPath = new File("train/");
        FileSplit fileSplitTrain = new FileSplit(mainPath, NativeImageLoader.ALLOWED_FORMATS, rng);
        mainPath = new File("test/");
        FileSplit fileSplitTest = new FileSplit(mainPath, NativeImageLoader.ALLOWED_FORMATS, rng);
        DataNormalization scaler = new ImagePreProcessingScaler(0, 1);

        log.info("Build model....");
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations) // Training iterations as above
                .regularization(true).l2(0.0005)
                .learningRate(.01)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .list()
                .layer(0, new ConvolutionLayer.Builder(5, 5)
                        .nIn(1)
                        .stride(1, 1)
                        .nOut(20)
                        .activation(Activation.IDENTITY)
                        .build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(2, new ConvolutionLayer.Builder(5, 5)
                        //Note that nIn need not be specified in later layers
                        .stride(1, 1)
                        .nOut(50)
                        .activation(Activation.IDENTITY)
                        .build())
                .layer(3, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(4, new DenseLayer.Builder().activation(Activation.RELU)
                                                  .nOut(500).build())
                .layer(5, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nOut(106)
                        .activation(Activation.SOFTMAX)
                        .build())
                .setInputType(InputType.convolutional(28, 28, 1)) //See note below
                .backprop(true).pretrain(false).build();
        MultiLayerNetwork network = new MultiLayerNetwork(conf);
        network.init();

        ImageRecordReader recordReader = new ImageRecordReader(height, width, channels, labelMaker);
        DataSetIterator dataIter;

        log.info("Train model....");
        // Train without transformations
        recordReader.initialize(fileSplitTrain, null);
        dataIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, numLabels);
        scaler.fit(dataIter);
        dataIter.setPreProcessor(scaler);
        network.fit(dataIter);

        log.info("Evaluate model....");
        ImageRecordReader recordReaderTest = new ImageRecordReader(height, width, channels, labelMaker);
        recordReaderTest.initialize(fileSplitTest);
        DataSetIterator dataIterTest = new RecordReaderDataSetIterator(recordReaderTest, batchSize, 1, numLabels);
        scaler.fit(dataIterTest);
        dataIterTest.setPreProcessor(scaler);
        Evaluation eval = new Evaluation(106);
        while(dataIterTest.hasNext()){
            DataSet ds = dataIterTest.next();
            INDArray output = network.output(ds.getFeatureMatrix(), false);
            eval.eval(ds.getLabels(), output);

        }
        log.info(eval.stats());
    }
`
Why my DataFetcher work better, then ImageRecordReader. What I do wrond?

**Comment by kobaeugenea** [comment]
There is a suspicion that the labels are incorrectly placed.

**Comment by agibsonccc** [comment]
I'm closing this. It's not properly worded. Please try maybe rewording what you're doing and try to describe your pipeline better?
